aodntools/timeseries_products/hourly_timeseries.py

```diff
@@ -0,0 +1,581 @@
+    t_end = pd.to_datetime(nc.TIME.max().values).strftime(file_timeformat)
+
+    output_name = '_'.join(
+        ['IMOS', facility_code, data_code, t_start, site_code, ('FV0' + str(file_version)), (VoI + "-" + product_type),
+         ('END-' + t_end), 'C-' + datetime.utcnow().strftime(file_timeformat)]) + '.nc'
+
+    return output_name
+
+
+
+def write_netCDF_aggfile(nc_aggregated, ncout_filename, encoding, file_path):
+    """
+    write netcdf file
+    :param file_path: path where to write the file
+    :param encoding: encoding dictionary
+    :param nc_aggregated: aggregated xarray dataset
+    :param ncout_filename: name of the netCDF file to be written
+    :return: name of the netCDf file written
+    """
+    ## sort the variables in the data set
+    variables_all = list(nc_aggregated.variables)
+    variables_head = ['instrument_index', 'instrument_id', 'source_file', 'TIME', 'LONGITUDE', 'LATITUDE',
+                      'NOMINAL_DEPTH', 'DEPTH', 'DEPTH_count', 'DEPTH_min', 'DEPTH_max', 'DEPTH_std', ]
+    variables_rest = sorted(list(set(variables_all) - set(variables_head)))
+    variables_all = variables_head + variables_rest
+
+    nc_aggregated[variables_all].to_netcdf(os.path.join(file_path, ncout_filename), encoding=encoding,
+                                           format='NETCDF4_CLASSIC')
+    return ncout_filename
+
+
+def append_aux_variables(filename, nc, df):
+    """
+    appends metadata variables to a dataframe
+
+    :param filename: str name of the source data file
+    :param nc: xarray dataset
+    :param df: pandas dataframe metadata variables
+    :return: pandas dataframe
+    """
+    df_temp = pd.DataFrame({'FILENAME': filename,
+                            'INSTRUMENT_TYPE': nc.attrs['deployment_code'] + '; ' + nc.attrs['instrument'] + '; ' +
+                                               nc.attrs['instrument_serial_number'],
+                            'LONGITUDE': nc.LONGITUDE.squeeze().values,
+                            'LATITUDE': nc.LATITUDE.squeeze().values,
+                            'NOMINAL_DEPTH': nc.NOMINAL_DEPTH.squeeze().values},
+                           index=[0])
+
+    return df.append(df_temp, sort=False, ignore_index=True)
+
+
+def PDresample_by_hour(df, function_dict, function_stats):
+    """
+    resample a dataframe by hour and calculate aggregation statistics
+    (mean, std, min, max, count)
+    the variables are renamed adding the corresponding suffix of the stats calculated
+    from the function_names list
+
+    :param function_stats: dictionary of binning ancillary stat functions
+    :param function_dict: dictionary of binning methods to be applied to each variable
+    :param df: pandas dataframe with ancillary variables and coords removed but with TIME as index
+    :return: pandas dataframe
+    """
+    ## back the index 30min
+    df.index = df.index - pd.Timedelta(30, units='m')
```

Comment:
Is this really the only way to get pandas to center the bins on the hour?

Reply:
Yes, as far as I've researched.

Reply:
see the[ function documentation,](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.resample.html), which is not very clear.